### ISSUE TITLE
Add hPutBuilder for Text's Builder type

### ISSUE THREAD

**Issue by axman6-da** [haskell/text#446]
Based on discussion in #442, having a way to write a text builder directly to a handle, like [bytestring's `hPutBuilder`](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#v:hPutBuilder) might be a nice performance win in some situations.

**Comment by nh2** [comment]
I would like to have a function with which I can write a Text Builder to a `Handle` while specifying the encoding, e.g. `hPutBuilderUtf8`.

For example,

```haskell
import System.IO
import qualified Data.Text.Lazy.IO as TL
main = withBinaryFile "test.txt" WriteMode $ \h -> TL.hPutStr h "Gebäude"
```

With `text` (at least version 1.*) this writes "garbage" into the file in the sense that it writes "truncated UTF-16`, which cannot be read by most other programs.

For small values like the above one could call `encodeUtf8` directly, but for large builder values one really wants to use an equivalent of `TL.hPutStr` which doesn't pick a non-useful encoding on `withBinaryFile`, and instead allows providing the encoding directly as an argument or in the function name.

**Comment by nh2** [comment]
I suspect that `Data.Text.Lazy.Encoding`'s [`encodeUtf8Builder :: Lazy.Text -> ByteString.Builder`](https://hackage.haskell.org/package/text-2.0.2/docs/Data-Text-Lazy-Encoding.html#v:encodeUtf8Builder) does the job:


```haskell
import Data.Text.Lazy.Encoding (encodeUtf8Builder)
import Data.ByteString.Builder (hPutBuilder)
import qualified Data.Text.Lazy.Builder as TB

withBinaryFile "test.txt" WriteMode $ \h -> do
  hPutBuilder h (encodeUtf8Builder $ TB.toLazyText $ TB.fromText "Gebäude")
```

**Comment by phadej** [comment]
`TB.toLazyText` + `encodeUtf8Builder` is unnecessary indirection. The `text`'s `Builder` is filling up buffers, these can be output directly to the `Handle` without allocating anything else in between (reusing the same buffer!)

**Comment by nh2** [comment]
@phadej Agreed. It would be nice to have a function that goes straight from a text `Builder` to syscalls.

**Comment by BebeSparkelSparkel** [comment]
Is there a way to just use [hPutBuilder](https://github.com/haskell/bytestring/blob/46a3aeb179c26d2385f95ba518eaa5464c94ceaa/Data/ByteString/Builder.hs#L293) or the internal function [hPut](https://github.com/haskell/bytestring/blob/46a3aeb179c26d2385f95ba518eaa5464c94ceaa/Data/ByteString/Builder/Internal.hs#L605) from bytestring.

**Comment by BebeSparkelSparkel** [comment]
@phadej I do not see how this is any different using the lazy text hPutStr since Builder can only produce `[StrictText]`.

**Comment by BebeSparkelSparkel** [comment]
It seems that for this to work well a few changes need to be made

- `ensureFree` should use a given buffer size similar to toLasyTextWith so that it can match the handles buffer size
- `newBuffer` should be able to use `new` or `newPinned` because newPinned will allow the handle to use the buffer without copying

**Comment by BebeSparkelSparkel** [comment]
@Lysxia You are correct with the current implementation of builder. However, in #586 I have modified Builder to be able to also be able to use pinned buffers. Pinned buffers allow the handle to use the same buffer that the builder uses thus skipping the copy from the builder buffer to the handle char buffer. This is helpful for all encodings that are not UTF-8.

**Comment by Lysxia** [comment]
I'm not convinced adding `hPutBuilder` is the right thing to do. If you want to append text values efficiently before printing them (which is where this issue originally comes from), you go through `ByteString.Builder` with [encodeUtf8Builder](https://hackage.haskell.org/package/text-2.0.2/docs/Data-Text-Encoding.html#v:encodeUtf8Builder). `Text.Lazy.Builder` is the wrong data structure for this (and it's specialized to UTF-8, whereas one could implement other encoders to `ByteString.Builder`), and since there is already a suitable data structure elsewhere, there's no point changing it.

**Comment by BebeSparkelSparkel** [comment]
I see what you mean. Thanks

**Comment by BebeSparkelSparkel** [comment]
Seems like this should be closed.
